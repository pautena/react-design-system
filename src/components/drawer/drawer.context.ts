import {useContext,createContext} from 'react';


export interface DrawerContextProps {
  isOpen: boolean;
  close(): void;
  open(): void;
}

export const DrawerContext = createContext<DrawerContextProps | undefined>(undefined);

export const useDrawer = ()=> {
  const context = useContext(DrawerContext);

  if (context === undefined) {
    throw new Error("DrawerContext.Provider is required and was undefined");
  }

  return context;
}