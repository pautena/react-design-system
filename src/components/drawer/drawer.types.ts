import { FunctionComponent, ReactElement } from "react";
import { BulletVariant } from "../bullet";
import { LabelVariant } from "../label";
import {
  DrawerProps as MuiDrawerProps,
} from "@mui/material";


export interface NavItem {
  id: string;
  name: string;
  icon?: ReactElement;
  avatar?:{
    src: string;
    alt: string;
  };
  label?:{
    text:string;
    variant:LabelVariant;
  };
  bullet?: {
    variant: BulletVariant;
  }
}

export interface NavSection {
  title?: string;
  items:  NavItem[];
}

export interface  Nav {
  items: NavSection[];
};

export interface DrawerContentProps {
  nav: Nav;
  open:boolean;
}
export type DrawerContentComponent = FunctionComponent<DrawerContentProps>;
export type DrawerContentElement = ReactElement<DrawerContentProps,DrawerContentComponent>;

export interface DrawerProps extends MuiDrawerProps {
  open: boolean;
  onDrawerClose: () => void;
  children: DrawerContentElement;
}

export type DrawerComponent = FunctionComponent<DrawerProps>;
export type DrawerElement = ReactElement<DrawerProps,DrawerComponent>;


