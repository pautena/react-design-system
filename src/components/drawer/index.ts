export * from "./drawer";
export * from './drawer.types'
export * from './drawer-content'
export * from './mini-drawer';
export * from './drawer.provider';
export * from './drawer.context'