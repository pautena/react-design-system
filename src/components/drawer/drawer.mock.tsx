import React from "react";
import { Nav } from "./drawer.types";
import DiamondIcon from "@mui/icons-material/Diamond";
import MenuBookIcon from "@mui/icons-material/MenuBook";

export const mockNav: Nav = {
  items: [
    {
      items: [
        {
          id: "item1.1",
          text: "Item 1.1",
          icon: <DiamondIcon />,
          href: "/items/1-1",
        },
        {
          id: "item1.2",
          text: "Item 1.2",
          href: "/items/1-2",
          icon: <DiamondIcon />,
        },
        {
          id: "item1.3",
          text: "Item 1.3",
          href: "/items/1-3",
          icon: <DiamondIcon />,
          bullet: {
            variant: "secondary",
          },
        },
        {
          id: "item1.4",
          text: "Item 1.4",
          href: "/items/1-4",
          icon: <DiamondIcon />,
          bullet: {
            variant: "error",
          },
        },
      ],
    },
    {
      title: "Section 2",
      items: [
        {
          id: "item2.1",
          text: "Item 2.1",
          href: "/items/2-1",
          icon: <DiamondIcon />,
          label: {
            text: "10",
            variant: "error",
          },
        },
        {
          id: "item2.2",
          text: "Item 2.2",
          href: "/items/2-2",
          icon: <MenuBookIcon />,
          label: {
            text: "new",
            variant: "info",
          },
        },
      ],
    },
    {
      title: "Section avatar",
      items: [
        {
          id: "item.a.1",
          text: "Avatar 1",
          href: "/items/a-1",
          avatar: {
            src: "https://c.pxhere.com/images/f6/e0/022ab77c391925931fe227c97ab4-1447675.jpg!d",
            alt: "Avatar 1",
          },
        },
        {
          id: "item.a.2",
          text: "Avatar 2",
          href: "/items/a-2",
          avatar: {
            src: "https://www.publicdomainpictures.net/pictures/270000/velka/avatar-people-person-business-u-15354603894rE.jpg",
            alt: "Avatar 2",
          },
        },
      ],
    },
  ],
};
