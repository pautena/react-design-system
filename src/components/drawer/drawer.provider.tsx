import {useState} from 'react';
import React,{ PropsWithChildren } from "react"
import {DrawerContext} from "./drawer.context";

type Props = PropsWithChildren<{ 
  initialOpen?:boolean;
}>

export const DrawerProvider = ({children,initialOpen=false}:Props) => {
  const [open,setOpen] = useState(initialOpen);

  return (
    <DrawerContext.Provider value={{
      isOpen:open,
      close: ()=> setOpen(false),
      open: ()=> setOpen(true),
    }}>
      {children}
    </DrawerContext.Provider>
  );
}