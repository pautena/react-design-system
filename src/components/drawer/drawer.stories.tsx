import React from "react";
import { Meta, StoryObj } from "@storybook/react";
import { withMemoryRouter } from "~/storybook";
import { DrawerContent } from "../drawer-content";
import { Drawer, DrawerProvider } from "../drawer-provider";
import { mockDrawerNavigation } from "../drawer.mock";
import { DrawerState } from "../drawer.types";

interface DrawerDemoArgs {
  initialState: DrawerState;
  selectedItemId?: string;
}

export default {
  title: "Components/Drawers/Drawer",
  decorators: [withMemoryRouter()],
  parameters: {
    layout: "fullscreen",
  },
  render: ({ initialState, selectedItemId }) => (
    <DrawerProvider initialState={initialState} variant="mini" selectedItemId={selectedItemId}>
      <Drawer>
        <DrawerContent nav={mockDrawerNavigation} />
      </Drawer>
    </DrawerProvider>
  ),
} satisfies Meta<DrawerDemoArgs>;
type Story = StoryObj<DrawerDemoArgs>;

export const Default: Story = {
  args: {
    selectedItemId: "item2.3.2",
    initialState: "open",
  },
};
