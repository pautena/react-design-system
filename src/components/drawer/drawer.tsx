import React from "react";
import { styled, useTheme } from "@mui/material/styles";
import { Drawer as MuiDrawer, Divider, IconButton } from "@mui/material";
import ChevronLeftIcon from "@mui/icons-material/ChevronLeft";
import { DrawerComponent, DrawerProps } from "./drawer.types";
import { useDrawer } from "./drawer.context";

const DrawerHeader = styled("div")(({ theme }) => ({
  display: "flex",
  alignItems: "center",
  justifyContent: "flex-end",
  padding: theme.spacing(0, 1),
  ...theme.mixins.toolbar,
}));

export const Drawer: DrawerComponent = ({ children, ...rest }: DrawerProps) => {
  const theme = useTheme();
  const { isOpen, close } = useDrawer();

  return (
    <MuiDrawer open={isOpen} role="menu" aria-hidden={!isOpen} {...rest}>
      <DrawerHeader>
        <IconButton onClick={close}>
          <ChevronLeftIcon />
        </IconButton>
      </DrawerHeader>
      <Divider />
      {children}
    </MuiDrawer>
  );
};
