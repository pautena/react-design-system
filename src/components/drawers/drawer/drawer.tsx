import React from "react";
import {
  SxProps,
  Theme,
  drawerClasses,
  styled,
  Drawer as MuiDrawer,
  Divider,
  IconButton,
  useTheme,
  paperClasses,
} from "@mui/material";
import ChevronLeftIcon from "@mui/icons-material/ChevronLeft";
import { DrawerComponent, DrawerProps, DrawerState, DrawerVariant } from "../drawer.types";
import { useDrawer } from "../drawer-provider/drawer-context";
import { closedMixin, openedMixin } from "../drawer-provider/drawer-mixins";

export const DrawerHeader = styled("div")(({ theme }) => ({
  display: "flex",
  alignItems: "center",
  justifyContent: "flex-end",
  padding: theme.spacing(0, 1),
  ...theme.mixins.toolbar,
}));

const showCloseButton: Record<DrawerVariant, boolean> = {
  temporary: true,
  mini: true,
  persistent: true,
  clipped: false,
};

const muiDrawerVariant: Record<DrawerVariant, "permanent" | "persistent" | "temporary"> = {
  temporary: "temporary",
  mini: "permanent",
  clipped: "permanent",
  persistent: "persistent",
};

type SxGenerator = (state: DrawerState, theme: Theme) => SxProps<Theme>;
const NoopSxGenerator = () => ({});
const variantsSx: Readonly<Record<DrawerVariant, SxGenerator>> = {
  mini: (state: DrawerState, theme: Theme) => ({
    boxSizing: "border-box",
    [`& .${paperClasses.root}`]: {
      zIndex: theme.zIndex.drawer - 1,
    },
  }),
  temporary: NoopSxGenerator,
  clipped: NoopSxGenerator,
  persistent: NoopSxGenerator,
};

export const Drawer: DrawerComponent = ({ children, ...rest }: DrawerProps) => {
  const theme = useTheme();
  const { state, switchState, underAppBar, close, drawerWidth, variant } = useDrawer();

  const sx: any = {
    width: drawerWidth,
    flexShrink: 0,
    whiteSpace: "nowrap",
    ...(state === "open" && {
      ...openedMixin(theme),
      [`& .${drawerClasses.paper}`]: openedMixin(theme),
    }),
    ...(state !== "open" && {
      ...closedMixin(theme),
      [`& .${drawerClasses.paper}`]: closedMixin(theme),
    }),
    ...variantsSx[variant](state, theme),
  };

  return (
    <MuiDrawer
      open={state === "open"}
      variant={muiDrawerVariant[variant]}
      role="menu"
      aria-hidden={state === "close"}
      onClose={close}
      sx={sx}
      {...rest}
    >
      <DrawerHeader>
        {!underAppBar && showCloseButton[variant] && (
          <IconButton onClick={switchState}>
            <ChevronLeftIcon />
          </IconButton>
        )}
      </DrawerHeader>
      <Divider />
      {children}
    </MuiDrawer>
  );
};
