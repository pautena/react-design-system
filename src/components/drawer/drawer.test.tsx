import React from "react";
import { Drawer } from "./drawer";
import { render, screen } from "../../tests";
import { DrawerProvider } from "./drawer.provider";
import { DummyDrawerContent, mockNav } from "./drawer.mock";
import { Box, Button, createTheme } from "@mui/material";
import { useDrawer } from "./drawer.context";
import userEvent from "@testing-library/user-event";
import {openedMixin,closedMixin} from './drawer.mixins'

describe("Drawer", () => {
  const renderComponent = ({ initialOpen = undefined }: { initialOpen?: boolean } = {}) => {
    const TestContent = () => {
      const { open, close } = useDrawer();

      return (
        <Box>
          <Button onClick={close}>close</Button>
          <Button onClick={open}>open</Button>
        </Box>
      );
    };
    return render(
      <DrawerProvider initialOpen={initialOpen}>
        <Box>
          <Drawer>
            <DummyDrawerContent nav={mockNav} />
          </Drawer>
          <TestContent />
        </Box>
      </DrawerProvider>,
    );
  };

  it("would render closed by default", () => {
    renderComponent();

    expect(screen.queryByRole("menu", { hidden: true })).not.toBeInTheDocument();
  });

  it("would render opened if initialOpen is true", () => {
    renderComponent({ initialOpen: true });

    expect(screen.getByRole("menu", { hidden: false })).toBeInTheDocument();
  });

  it("would open if the user click the open button", async () => {
    renderComponent();

    await userEvent.click(screen.getByText(/open/i));

    expect(screen.getByRole("menu", { hidden: false })).toBeInTheDocument();
  });

  it("would close if the user click the close button and the drawer was opened", async () => {
    renderComponent({ initialOpen: true });

    await userEvent.click(screen.getByText(/close/i));

    expect(screen.queryByRole("menu", { hidden: true })).not.toBeInTheDocument();
  });
});

describe("mixins",()=>{
  describe('openedMixin',()=>{
    it('would match a snapshot',()=>{
      const theme = createTheme();
      expect(openedMixin(theme)).toMatchSnapshot();
    })
  })

  describe('closedMixin',()=>{
    it('would match a snapshot',()=>{
      const theme = createTheme();
      expect(closedMixin(theme)).toMatchSnapshot();
    })
  })
})
