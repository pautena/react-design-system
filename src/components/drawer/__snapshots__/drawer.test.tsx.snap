// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mixins closedMixin would match a snapshot 1`] = `
Object {
  "@media (min-width:600px)": Object {
    "width": "calc(64px + 1px)",
  },
  "overflowX": "hidden",
  "transition": "width 195ms cubic-bezier(0.4, 0, 0.6, 1) 0ms",
  "width": "calc(56px + 1px)",
}
`;

exports[`mixins openedMixin would match a snapshot 1`] = `
Object {
  "overflowX": "hidden",
  "transition": "width 225ms cubic-bezier(0.4, 0, 0.6, 1) 0ms",
  "width": 240,
}
`;
